### DISCUSSION TITLE
Restart Detector

### DISCUSSION THREAD

**Discussion by manofst3el72** [MMRIZE/MMM-Hotword2#1] [Q&A]
It doesn't happen right away, but I will get an error "pvrecorder failed to read audio data frame"

My question is about restarting the detector when it errors.  I can restart my mirror and then it works fine for a while but eventually the error returns.  

I tried adjusting the soundThreshold thinking it was something the microphone was picking up causing the error but I couldn't get the error to stop,

My thoughts are just restart the detector but if the error could be resolved another way that would work as well

**Comment by eouia** [comment]
I’ll consider

**Comment by eouia** [comment]
Can you provide the exact error log when it happened?

**Reply by manofst3el72** [reply]
Does this help?


```
0|MagicMirror  | [2024-09-15 11:19:49.038] [LOG]   [HOT2] All previously recorded files are deleted.'
0|MagicMirror  | [2024-09-15 11:19:49.060] [LOG]   [HOT2] Hotword added: '
'MISTER MIRROR Mister-Mirror_en_raspberry-pi_v3_0_0.ppn'
0|MagicMirror  | [2024-09-15 11:19:52.985] [LOG]   [HOT2] MIC device: PCM2902 Audio Codec Analog Mono
0|MagicMirror  | [2024-09-15 11:19:52.988] [LOG]   [HOT2] Detector started
0|MagicMirror  | [2024-09-15 11:19:52.993] [INFO]  updatenotification: Updater Class Loaded!
0|MagicMirror  | [2024-09-15 11:19:52.994] [INFO]  updatenotification: Checking PM2 using...
0|MagicMirror  | [2024-09-15 11:19:56.553] [INFO]  Checking git for module: MMM-Hotword2
0|MagicMirror  | [2024-09-15 11:19:58.314] [INFO]  [MMM-DynamicWeather] Received successful Weather API data
0|MagicMirror  | [2024-09-15 11:19:58.547] [INFO]  Checking git for module: MMM-Sounds
....
0|MagicMirror  | [2024-09-15 11:29:37.429] [INFO]  [MMM-DynamicWeather] Getting Weather API data
0|MagicMirror  | [2024-09-15 11:29:39.917] [INFO]  [MMM-DynamicWeather] Received successful Weather API data
0|MagicMirror  | [2024-09-15 11:39:37.401] [INFO]  [MMM-DynamicWeather] Getting Weather API data
0|MagicMirror  | [2024-09-15 11:39:39.913] [INFO]  [MMM-DynamicWeather] Received successful Weather API data
0|MagicMirror  | [2024-09-15 11:42:57.584] [LOG]   [HOT2] Detector process failed.
0|MagicMirror  | [2024-09-15 11:42:57.971] [LOG]   [HOT2] PvRecorderStatusIOError: PvRecorder failed to read audio data frame.
0|MagicMirror  |     at pvRecorderStatusToException (/home/smartmirror/MagicMirror/modules/MMM-Hotword2/node_modules/@picovoice/pvrecorder-node/dist/errors.js:45:20)
0|MagicMirror  |     at Timeout._onTimeout (/home/smartmirror/MagicMirror/modules/MMM-Hotword2/node_modules/@picovoice/pvrecorder-node/dist/pv_recorder.js:107:49)
0|MagicMirror  |     at listOnTimeout (node:internal/timers:581:17)
0|MagicMirror  |     at process.processTimers (node:internal/timers:519:7)`
```

**Comment by eouia** [comment]
I suspect one thing: MMM-Sounds or other mic/speaker-related modules may interfere with the system.
To make things simple, could you test again without other unrelated modules? dropout other modules or comment-out or with `disabled: true`

**Reply by manofst3el72** [reply]
Getting the same thing with MMM-Sounds commented out

```
0|MagicMirror  | [2024-09-16 20:01:46.858] [INFO]  updatenotification: Checking PM2 using...
0|MagicMirror  | [2024-09-16 20:01:49.431] [INFO]  Checking git for module: MMM-Hotword2
0|MagicMirror  | [2024-09-16 20:01:49.582] [LOG]   [HOT2] All previously recorded files are deleted.
0|MagicMirror  | [2024-09-16 20:01:49.596] [LOG]   [HOT2] Hotword added: MISTER MIRROR Mister-Mirror_en_raspberry-pi_v3_0_0.ppn
0|MagicMirror  | [2024-09-16 20:01:53.911] [LOG]   [HOT2] MIC device: PCM2902 Audio Codec Analog Mono
0|MagicMirror  | [2024-09-16 20:01:53.920] [LOG]   [HOT2] Detector started
...
0|MagicMirror  | [2024-09-16 20:01:55.061] [INFO]  Checking git for module: MMM-pages
0|MagicMirror  | [2024-09-16 20:01:55.284] [INFO]  Checking git for module: MMM-page-indicator
0|MagicMirror  | [2024-09-16 20:01:55.877] [INFO]  Checking git for module: MMM-DynamicWeather
0|MagicMirror  | [2024-09-16 20:01:59.258] [INFO]  updatenotification: [PM2] You are not using pm2
0|MagicMirror  | [2024-09-16 20:01:59.276] [INFO]  Checking git for module: MMM-BurnIn
0|MagicMirror  | [2024-09-16 20:02:00.878] [INFO]  Checking git for module: MMM-NewsFeed
0|MagicMirror  | [2024-09-16 20:02:01.861] [INFO]  Checking git for module: MMM-NewsAPI
0|MagicMirror  | [2024-09-16 20:02:03.016] [INFO]  Checking git for module: MMM-Traffic
0|MagicMirror  | [2024-09-16 20:02:03.271] [INFO]  Checking git for module: MagicMirror
0|MagicMirror  | [2024-09-16 20:11:44.002] [INFO]  [MMM-DynamicWeather] Getting Weather API data
0|MagicMirror  | [2024-09-16 20:11:44.563] [INFO]  [MMM-DynamicWeather] Received successful Weather API data
0|MagicMirror  | [2024-09-16 20:21:44.386] [INFO]  [MMM-DynamicWeather] Getting Weather API data
0|MagicMirror  | [2024-09-16 20:21:49.973] [INFO]  [MMM-DynamicWeather] Received successful Weather API data
0|MagicMirror  | [2024-09-16 20:31:44.130] [INFO]  [MMM-DynamicWeather] Getting Weather API data
0|MagicMirror  | [2024-09-16 20:31:44.670] [INFO]  [MMM-DynamicWeather] Received successful Weather API data
0|MagicMirror  | [2024-09-16 20:41:44.435] [INFO]  [MMM-DynamicWeather] Getting Weather API data
0|MagicMirror  | [2024-09-16 20:41:48.775] [INFO]  [MMM-DynamicWeather] Received successful Weather API data
0|MagicMirror  | [2024-09-16 20:42:31.978] [LOG]   [HOT2] Detector process failed.
0|MagicMirror  | [2024-09-16 20:42:32.859] [LOG]   [HOT2] PvRecorderStatusIOError: PvRecorder failed to read audio data frame.
0|MagicMirror  |     at pvRecorderStatusToException (/home/smartmirror/MagicMirror/modules/MMM-Hotword2/node_modules/@picovoice/pvrecorder-node/dist/errors.js:45:20)
0|MagicMirror  |     at Timeout._onTimeout (/home/smartmirror/MagicMirror/modules/MMM-Hotword2/node_modules/@picovoice/pvrecorder-node/dist/pv_recorder.js:107:49)
0|MagicMirror  |     at listOnTimeout (node:internal/timers:581:17)
0|MagicMirror  |     at process.processTimers (node:internal/timers:519:7)
0|MagicMirror  | [2024-09-16 20:51:44.218] [INFO]  [MMM-DynamicWeather] Getting Weather API data
0|MagicMirror  | [2024-09-16 20:51:44.725] [INFO]  [MMM-DynamicWeather] Received successful Weather API data
0|MagicMirror  | [2024-09-16 21:01:45.820] [INFO]  [MMM-DynamicWeather] Getting Weather API data
0|MagicMirror  | [2024-09-16 21:01:46.954] [ERROR] [MMM-DynamicWeather] Failed getting api:  AggregateError [ETIMEDOUT]:
```
'''

**Reply by eouia** [reply]
Thanks for confirming

**Comment by eouia** [comment]
First, if I don't know exactly what caused the error, restarting it won't solve the problem. The problem will occur again; this time, it will go into an infinite loop because of the restart.

I have a few more questions to clarify the symptoms.
- Tell me what hardware (Device, Mic), OS, and NodeJS version you use. It looks like you are using a Raspberry Pi.
- Show me the Hotword2 config you are currently using.
- Does the error always occur? Or sometimes it occurs, and sometimes it is fine?
- Looking at the log, it seems like the error occurs right after starting. Is there a case where it suddenly occurs after normal operation? (For example, sometimes the detector works properly until the 2nd time but stops at the 3rd.)
- This error usually occurs when data (frames) cannot be read properly from the microphone. It can be a hardware problem, or a software problem. Did you change the `recorderFrameLength` or `recorderBufferedFramesCount` properties?


[Test More]
- To simplify the diagnosis, disable all other modules except MMM-Hotword2.
- Change the config of MMM-Hotword2 to the simplest form (see below) and test it.
```js
{
	module: "MMM-Hotword2",
	position: "bottom_right",
	config: {
		verbose: true,
		hotwords: [
			{
				hotword: "COMPUTER",
			}
		]
	}
}
```

**Reply by manofst3el72** [reply]
> Tell me what hardware (Device, Mic), OS, and NodeJS version you use. It looks like you are using a Raspberry Pi.
RASPBERRY PI 3 MOD A+ 
USB Microphone
Raspbian GNU/Linux 12 (bookworm) 
node version 20.17
> Show me the Hotword2 config you are currently using.
```
	{
	module: "MMM-Hotword2",
	position: "bottom_left",
	startOnBoot: true,
	config: {
	  verbose: true,
	  device: 2,
	  soundThreshold: 0.6,
	  hotwords: [
		{ hotword: 'MISTER MIRROR',
			file: "Mister-Mirror_en_raspberry-pi_v3_0_0.ppn",
		continuousRecording: false,
		restart:true,
		onDetect:  ({ helper, error, result, payload }) => 
		{
		
			console.log("Detected hotword: 'Mister Mirror'")
			helper.sendNotification("PLAY_SOUND","weather.wav")
			helper.sendNotification("SHOW_HIDDEN_PAGE","weather")
			
			setTimeout(() => helper.sendNotification("LEAVE_HIDDEN_PAGE"), 60000)
		},
		}
	  ]},
		},
```
> Does the error always occur? Or sometimes it occurs, and sometimes it is fine?

It starts out fine and will detect the hotword but after 30 minutes to an hour the error shows up

> Looking at the log, it seems like the error occurs right after starting. Is there a case where it suddenly occurs after normal operation? (For example, sometimes the detector works properly until the 2nd time but stops at the 3rd.)

Saying the hotword and being detected seems to have no effect.  The hotword can be detected then it goes back to waiting but then eventually the error occurs

> This error usually occurs when data (frames) cannot be read properly from the microphone. It can be a hardware problem, or a software problem. Did you change the recorderFrameLength or recorderBufferedFramesCount properties? 

I tried changing those with the config option but it did not seem to help so I removed them from my config



I will report back about how it works with less modules and start adding them back one at a time

**Reply by eouia** [reply]
Does the error occur 1) when the detector is in standby mode for a while? 2) Or does the error occur immediately (within a few seconds) after hotword detection? (I can't tell because there is no log related to the previous hotword detection in the log you provided.)

If it's the first case, there's no way to prevent it in advance, and a practical solution might be to automatically restart when the error occurs.

If it's the second case, I guess it's because the Raspberry Pi 3 is a slow device, so it's trying to open the microphone before the microphone connection from the previous detection is closed really.

The problem is that it's difficult to check when the microphone closing is finished synchronously.

If `continuousRecording: true`, the detector is restarted immediately when `onDetect` is executed, so to solve this problem,

- If `continuousRecording: true`, give a small delay before re-execution; I'll make an update if the guess is right.

- Since `onDetect` is an `async` function, it would be possible to terminate it after a certain amount of delay within onDetect.

**Reply by eouia** [reply]
> * Since `onDetect` is an `async` function, it would be possible to terminate it after a certain amount of delay within onDetect.

```javascript
onDetect: async ({ helper }) => {
	const asleep = (ms) => new Promise(resolve => setTimeout(resolve, ms))	
	console.log("Detected hotword: 'Mister Mirror'")
	helper.sendNotification("PLAY_SOUND","weather.wav")
	helper.sendNotification("SHOW_HIDDEN_PAGE","weather")
	await asleep(60000)
	helper.sendNotification("LEAVE_HIDDEN_PAGE")
	// After LEAVE_HIDDEN_PAGE, the hotword detection will be re-enabled
},
```

**Reply by manofst3el72** [reply]
I have no problems until I add the following:

MMM-NewsAPI
MMM-NewsFeed
MMM-Traffic

I suspect it is something that happens when data is pulled for these.  

Will terminating it after a certain amount of delay work without the word being detected?

**Reply by eouia** [reply]
First, this kind of problem is very hard to reproduce, so the exact cause isn't easy to pinpoint.

Based on your symptoms, a new detection cycle(right after `onDetect()` is executed, because of `restart: true`) seems to be starting before the microphone is completely closed. There could be several reasons for this. Your computer may be slow for a particular reason, or other processes may be using the microphone.
In any case, since the microphone closing is not synchronized, it might be worth adding a small delay at the end of one cycle just to be safe.
The `onDetect` function is an `async` function. After this function is executed, the new cycle will start. So you can control the cycle by yourself.
```javascript
onDetect: async ({ helper, error }) => {
  const asleep = (ms) => new Promise(resolve => setTimeout(resolve, ms))	
    ... // Your jobs
  await asleep(3000) // <- wait 3 seconds before the end of `onDetect`
}
```
I think 3 seconds would be enough to close the mic. 

> Will terminating it after a certain amount of delay work without the word being detected?

Maybe I can implement that feature, but in that case, after sleep, how can you reactivate this module?

I'm considering introducing an `onError` callback to handle this kind of error, but without guessing the exact causes, I'm worrying about it couldn't be a final solution.

**Reply by eouia** [reply]
One more thing; If possible, can you show the front-end log also? (You showed me the backend log, but there is also browser-level log. You can check it with `Chrome dev-console` or `MMM-LogExt` module.)

**Reply by manofst3el72** [reply]
I found this in the front end log

```
Uncaught (in promise) TypeError: Cannot destructure property 'sender' of 'this.consumeRequestedMap(...)' as it is null.
    at Class.onResult (MMM-Hotword2.js:115:13)
    at Class.socketNotificationReceived (MMM-Hotword2.js:182:34)
    at module.js:231:9
    at Socket.<anonymous> (socketclient.js:32:4)
    at Emitter.emit (index.mjs:136:20)
    at Socket.emitEvent (socket.js:538:20)
    at Socket.onevent (socket.js:525:18)
    at MMSocket.socket.onevent (socketclient.js:26:11)
    at Socket.onpacket (socket.js:495:22)
    at Emitter.emit (index.mjs:136:20)
```

**Reply by eouia** [reply]
Thank you

**Comment by manofst3el72** [comment]
The error occurs when the detector is in standby mode for a while.

**Comment by Montanadude** [comment]
> It doesn't happen right away, but I will get an error "pvrecorder failed to read audio data frame"
> 
> My question is about restarting the detector when it errors. I can restart my mirror and then it works fine for a while but eventually the error returns.
> 
> I tried adjusting the soundThreshold thinking it was something the microphone was picking up causing the error but I couldn't get the error to stop,
> 
> My thoughts are just restart the detector but if the error could be resolved another way that would work as well

I would also like to see this.   I am running this on a PI4 compute and I also see the same thing and restarting the mirror is what fixes the issue.

**Reply by eouia** [reply]
Sorry for your guys' inconvenience. I'm testing the new version, but as you know, this kind of issue is not so easy to reproduce intentionally, so I do not have enough certainty to release it.
